tiledb/sm/c_api/tiledb.cc

```diff
@@ -1482,41 +1469,30 @@ capi_return_t tiledb_array_delete_fragments_v2(
   }
 
   // Allocate an array object
-  tiledb_array_t* array = new (std::nothrow) tiledb_array_t;
-  try {
-    array->array_ =
-        make_shared<tiledb::sm::Array>(HERE(), ctx->resources(), uri);
-  } catch (...) {
-    delete array;
-    array = nullptr;
-    throw api::CAPIStatusException("Failed to create array");
-  }
+  tiledb::sm::Array array(ctx->resources(), uri);
 
   // Set array open timestamps
-  array->array_->set_timestamp_start(timestamp_start);
-  array->array_->set_timestamp_end(timestamp_end);
+  array.set_timestamp_start(timestamp_start);
+  array.set_timestamp_end(timestamp_end);
 
   // Open the array for exclusive modification
-  throw_if_not_ok(array->array_->open(
+  throw_if_not_ok(array.open(
       static_cast<tiledb::sm::QueryType>(TILEDB_MODIFY_EXCLUSIVE),
       static_cast<tiledb::sm::EncryptionType>(TILEDB_NO_ENCRYPTION),
       nullptr,
       0));
 
   // Delete fragments
   try {
-    array->array_->delete_fragments(uri, timestamp_start, timestamp_end);
+    array.delete_fragments(uri, timestamp_start, timestamp_end);
   } catch (...) {
-    throw_if_not_ok(array->array_->close());
-    delete array;
-    array = nullptr;
-    throw api::CAPIStatusException("Failed to delete fragments");
+    throw_if_not_ok(array.close());
+    std::throw_with_nested(
+        api::CAPIStatusException("Failed to delete fragments"));
   }
 
   // Close and delete the array
-  throw_if_not_ok(array->array_->close());
-  delete array;
-  array = nullptr;
+  throw_if_not_ok(array.close());
```

Comment:
I was tempted to just delete the `close` calls; the only thing they do is persist updated metadata to storage, which don't exist here. I would like us to one day formally document when `close` can be skipped (currently always, except for when metadata and/or group members in case of groups were changed).

Reply:
> formally document when `close` can be skipped

That's easy: Never.

Anything else is allowing implementation details to bleed through to actual use. `class Array` should be able to rely on matched open/close pairs in user code. 

The behavior of not writing when an array is not closed before it's destroyed is a poor design. `class Array` from the perspective of a transactional database is something like an RM (resource manager, a term used in transactions differently than core typically uses it). The transactional pattern would have an `abort()` function to discard any pending changes. For simplified, single-resource transactions, closing a resource with pending modifications has an implicit `commit` operation.

All that said, implicit operations on object destruction are a bad idea generally. C++ does not deal well with destructors that throw exceptions, but failing to close or abort an array should result in a user-visible error. The current state of the C API and the core implementation make that problematic, however desirable.

Reply:
> a transactional database

But we are not currently a transactional database, are we? At least in the sense relational databases are, considering at least that the storage layer is non-transactional across multiple files, and readers and writers cannot directly communicate with each other. You do not open and close objects on S3 and Azure. Design choices made by transactional databases are not necessarily the best for our case.

If we decide in the future to support more fully-featured transactions we can revisit this and add appropriate APIs (like a dedicated transaction object that can be attached to multiple arrays).

> implicit operations on object destruction are a bad idea generally. C++ does not deal well with destructors that throw exceptions

That's exactly why I don't like the existing design of `close`.